`EXPLAIN ANALYSE` should honor profiler output format

bleskes opened the pull request:
Today, the `EXPLAIN ANALYSE` statement always results in a query tree output, regardless of the profiler output format pragma. This commit makes it honor that setting.

To do so, the default output format had to be changed from `NONE` to `QUERY_TREE`, to avoid output from being suppressed by default. Note that we have another boolean that controls whether the profiler is on or not.

Closes #3094

bleskes committed:
`EXPLAIN ANALYSE` should honor profiler output format

Today, the `EXPLAIN ANALYSE` statement always results in a query tree output, regardless of the profiler output format pragma. This commit makes it honor that setting.

To do so, the default output format had to be changed from `NONE` to `QUERY_TREE`, to avoid output from being suppressed by default. Note that we have another boolean that controls whether the profiler is on or not.

Closes #3094

Mytherin commented on src/main/query_profiler.cpp:
```diff
@@ -122,15 +123,7 @@ void QueryProfiler::EndQuery() {
 	auto automatic_print_format = GetPrintFormat();
 	// print or output the query profiling after termination, if this is enabled
 	if (automatic_print_format != ProfilerPrintFormat::NONE) {
```
Could we perhaps turn this if into `if (!is_explain_analyze)`, and remove the `ProfilerPrintFormat::NONE` option entirely?

Mytherin commented on src/main/query_profiler.cpp:
```diff
@@ -141,6 +134,22 @@ void QueryProfiler::EndQuery() {
 	}
 	this->is_explain_analyze = false;
 }
+string QueryProfiler::PrintAsString() const {
```
Perhaps we could turn this method into `ToString()`, and rename the `ToString()` method to `ToQueryTree()`? That might be more clear.

bleskes replied:
Sure. I wasn't sure of the semantics of ToString and whether you always want a human readable version. WIll change.

bleskes replied:
@Mytherin The ToString makes it optional to print out the optimizer output and also it prints more things, like total time. Next to that, it is a sibling to the `ToStream` method. Do you still feel you want this change? (i.e., rename this method to be the `ToString`, and rename current `ToString`, and probably its sibling, to something else)

Mytherin replied:
I think it makes sense for the `ToString` to take the print profiler method into account, so the rename makes sense to me. `ToStream` should then indeed also be renamed to `QueryTreeToStream` or so.

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Some minor requests:

bleskes committed:
remove ProfilerPrintFormat.None

bleskes committed:
apparently we run the profiler all the time in the background, when in DEBUG mode. With the removal of `ProfilerPrintFormat.NONE`, we need another mechanism to suppress the output

bleskes committed:
Method renames

bleskes commented:
Thanks for the review @Mytherin . I addressed all your feedback and I think it's ready for another round. I took a little bit of an unconventional route to make sure we still suppress output when the profiler is silently turned on in debug mode so we don't have to change another debug-only flag. Hopefully you're ok with it.

Mytherin reviewed (commented):
Thanks! Looks good. One more minor comment, otherwise I think this is ready to merge:

bleskes commented on src/main/connection.cpp:
```diff
@@ -20,6 +20,7 @@ Connection::Connection(DatabaseInstance &database) : context(make_shared<ClientC
 	ConnectionManager::Get(database).AddConnection(*context);
 #ifdef DEBUG
 	EnableProfiling();
+	context->config.profiler_save_location = QueryProfiler::PROFILER_SHADOW_LOCATION;
```
k. I assume you mean a setting on the config (not the query profiler class, as you stated - I'm not sure how to get to it from the `PragmaEnableProfilingStatement` method). I'll work on that. Please shout if I misunderstood.

bleskes committed:
remove the magic string in favor of an explicit variable in config

bleskes commented:
@Mytherin I pushed another update, based on your feedback (and some assumptions). LMK what you think.

Mytherin commented on src/include/duckdb/main/client_config.hpp:
```diff
@@ -27,12 +27,18 @@ struct ClientConfig {
+#ifdef DEBUG
+	//! Allow suppressing profiler output, even if enabled. We turn on the profiler on all test runs but don't want
+	//! to output anything
+	bool emit_profiler_output = true;
```
We avoid having member variables' existence depend on defines, because this can cause nasty bugs where a component compiled with debug mode cannot interact with another component compiled in release mode which are very hard to track down in our experience.

Mytherin reviewed (commented):
Thanks for the changes, one more comment below. It also appears the regression tests are failing because of this change. Could you have a look at that?

bleskes committed:
remove ifdef DEBUG pragmas

bleskes committed:
do not enable profiler by default in benchmarks (triggered by it's output, now being visible)

bleskes committed:
Merge remote-tracking branch 'duckdb/master' into issue_3092

bleskes committed:
non interpreted benchmark code still had profiler enabled.
pragma for profiling mode didn't enable emitting.

bleskes committed:
formatting

Mytherin commented:
Thanks!

bleskes commented:
Thank you for the iterations, @Mytherin. Much appreciated
